ambari/src/main/java/io/brooklyn/ambari/AmbariClusterImpl.java

```diff
@@ -131,24 +106,28 @@ public void start(Collection<? extends Location> locations) {
 
         // TODO We could try to prevent setting SERVICE_UP until the services are all intalled
         subscribe(getAttribute(AMBARI_SERVER), AmbariServer.REGISTERED_HOSTS, new RegisteredHostEventListener(this));
+        Iterable<Entity> components = Entities.descendants(this, Predicates.instanceOf(SoftwareProcess.class));
+        EtcHostsManager.setHostsOnMachines(components, getConfig(ETC_HOST_ADDRESS));
     }
 
+
     static final class RegisteredHostEventListener implements SensorEventListener<List<String>> {
+
         private final AmbariCluster entity;
-        
+
         public RegisteredHostEventListener(AmbariCluster entity) {
             this.entity = entity;
         }
-        
         @Override
         public void onEvent(SensorEvent<List<String>> event) {
             List<String> hosts = event.getValue();
-            Integer initialClusterSize = entity.getConfig(INITIAL_SIZE);
+            Integer initialClusterSize = entity.getAttribute(EXPECTED_AGENTS);
             Boolean initialised = entity.getAttribute(CLUSTER_SERVICES_INITIALISE_CALLED);
             if (hosts != null && hosts.size() == initialClusterSize && !Boolean.TRUE.equals(initialised)) {
```

Comment:
installServices being called more than once probably wouldn't matter that much as ambari would reject the second attempt as the blueprint name will be unchanged.
However, @aledsage seemed pretty confident that the brooklyn thread model wouldn't allow this method to be called more than once concurrently.